File: knowledge_inferer/phase1_data_standardization.py
import pandas as pd
import numpy as np
import json
import os
from typing import Dict, Any, Optional


class DataStandardizer:
    """
    Phase 1: Data Standardization & Integration
    Chuẩn hóa và hợp nhất dữ liệu từ 3 nguồn báo cáo tài chính
    """
    
    def __init__(self, schema_mapping_path: str):
        with open(schema_mapping_path, 'r', encoding='utf-8') as f:
            self.schema = json.load(f)
        
        self.report_types = self.schema['report_types']
        self.merge_keys = self.schema['merge_keys']
        self.validation_fields = self.schema['validation_fields']
    
    def load_raw_data(self, raw_data_path: str) -> pd.DataFrame:
        """Bước 1: Đọc dữ liệu thô"""
        df = pd.read_csv(raw_data_path)
        print(f"✓ Loaded raw data: {df.shape[0]} rows, {df.shape[1]} columns")
        return df
    
    def split_by_report_type(self, df: pd.DataFrame) -> Dict[str, pd.DataFrame]:
        """Bước 2: Tách dữ liệu theo loại báo cáo"""
        reports = {}
        
        for report_type in self.report_types:
            reports[report_type] = df[df['report'] == report_type].copy()
            print(f"✓ {report_type}: {reports[report_type].shape[0]} rows")
        
        return reports
    
    def remove_empty_columns(self, reports: Dict[str, pd.DataFrame]) -> Dict[str, pd.DataFrame]:
        """Bước 3: Loại bỏ các cột toàn NaN"""
        cleaned = {}
        
        for report_type, df in reports.items():
            cleaned[report_type] = df.dropna(axis=1, how='all')
            removed = df.shape[1] - cleaned[report_type].shape[1]
            print(f"✓ {report_type}: Removed {removed} empty columns")
        
        return cleaned
    
    def merge_reports(self, reports: Dict[str, pd.DataFrame]) -> pd.DataFrame:
        """Bước 4: Hợp nhất 3 báo cáo"""
        keys = self.merge_keys
        
        balance = reports['balance_sheet']
        income = reports['income_statement'].drop(columns=['symbol'], errors='ignore')
        cash = reports['cash_flow'].drop(columns=['symbol'], errors='ignore')
        
        merged = (
            balance
            .merge(income, on=keys, how='left', validate='one_to_one')
            .merge(cash, on=keys, how='left', validate='one_to_one')
        )
        
        print(f"✓ Merged data: {merged.shape[0]} rows, {merged.shape[1]} columns")
        return merged
    
    def validate_data(self, df: pd.DataFrame) -> pd.DataFrame:
        """Bước 5: Kiểm tra tính hợp lệ của dữ liệu"""
        validation_rules = self.schema.get('validation_rules', [])
        
        for rule in validation_rules:
            rule_name = rule['rule']
            
            if rule_name == 'non_negative_assets':
                for field in rule['fields']:
                    col = self.validation_fields.get(field)
                    if col and col in df.columns:
                        negative_count = (df[col] < 0).sum()
                        if negative_count > 0:
                            print(f"⚠ Warning: {negative_count} negative values in {col}")
            
            elif rule_name == 'non_negative_liabilities':
                for field in rule['fields']:
                    col = self.validation_fields.get(field)
                    if col and col in df.columns:
                        negative_count = (df[col] < 0).sum()
                        if negative_count > 0:
                            print(f"⚠ Warning: {negative_count} negative values in {col}")
            
            elif rule_name == 'balance_equation':
                total_assets_col = self.validation_fields['total_assets']
                total_liabilities_col = self.validation_fields['total_liabilities']
                equity_col = self.validation_fields['owners_equity']
                
                if all(col in df.columns for col in [total_assets_col, total_liabilities_col, equity_col]):
                    tolerance = rule.get('tolerance', 0.01)
                    df['_balance_check'] = df[total_assets_col] - (df[total_liabilities_col] + df[equity_col])
                    violations = (df['_balance_check'].abs() > tolerance).sum()
                    df.drop(columns=['_balance_check'], inplace=True)
                    
                    if violations > 0:
                        print(f"⚠ Warning: {violations} rows violate balance equation")
        
        print(f"✓ Data validation completed")
        return df
    
    def standardize(self, raw_data_path: str) -> pd.DataFrame:
        """Thực hiện toàn bộ quy trình chuẩn hóa"""
        print("\n=== Phase 1: Data Standardization & Integration ===")
        
        df = self.load_raw_data(raw_data_path)
        reports = self.split_by_report_type(df)
        reports = self.remove_empty_columns(reports)
        merged = self.merge_reports(reports)
        validated = self.validate_data(merged)
        
        print("✓ Phase 1 completed\n")
        return validated
    
    def get_company_data(self, df: pd.DataFrame, ticker: str, year: int, quarter: int) -> Optional[pd.Series]:
        """Lấy dữ liệu của một công ty cụ thể"""
        ticker_col, year_col, quarter_col = self.merge_keys
        
        result = df[
            (df[ticker_col] == ticker) &
            (df[year_col] == year) &
            (df[quarter_col] == quarter)
        ]
        
        if result.empty:
            return None
        
        return result.iloc[0]

File: src/phases/Phase1DataStandardization.js
/**
 * Phase 1: Data Standardization & Integration
 * Chuẩn hóa và hợp nhất dữ liệu từ 3 nguồn báo cáo tài chính
 */

import fs from 'fs';
import csv from 'csv-parser';
import { createReadStream } from 'fs';

export class DataStandardizer {
  constructor(schemaMappingPath) {
    const schemaData = JSON.parse(fs.readFileSync(schemaMappingPath, 'utf-8'));
    this.schema = schemaData;
    this.reportTypes = schemaData.report_types;
    this.mergeKeys = schemaData.merge_keys;
    this.validationFields = schemaData.validation_fields;
  }

  /**
   * Bước 1: Đọc dữ liệu thô từ CSV
   */
  async loadRawData(rawDataPath) {
    return new Promise((resolve, reject) => {
      const data = [];
      
      createReadStream(rawDataPath)
        .pipe(csv())
        .on('data', (row) => {
          data.push(row);
        })
        .on('end', () => {
          resolve(data);
        })
        .on('error', (error) => {
          reject(error);
        });
    });
  }

  /**
   * Bước 2: Tách dữ liệu theo loại báo cáo
   */
  splitByReportType(data) {
    const reports = {};
    
    for (const reportType of this.reportTypes) {
      reports[reportType] = data.filter(row => row.report === reportType);
    }
    
    return reports;
  }

  /**
   * Bước 3: Loại bỏ các cột toàn null/undefined
   */
  removeEmptyColumns(reports) {
    const cleaned = {};
    
    for (const [reportType, rows] of Object.entries(reports)) {
      if (rows.length === 0) {
        cleaned[reportType] = rows;
        continue;
      }

      const allKeys = Object.keys(rows[0]);
      const nonEmptyKeys = allKeys.filter(key => {
        return rows.some(row => row[key] !== null && row[key] !== undefined && row[key] !== '');
      });

      cleaned[reportType] = rows.map(row => {
        const cleanedRow = {};
        for (const key of nonEmptyKeys) {
          cleanedRow[key] = row[key];
        }
        return cleanedRow;
      });

      const removed = allKeys.length - nonEmptyKeys.length;
    }
    
    return cleaned;
  }

  /**
   * Bước 4: Hợp nhất 3 báo cáo
   */
  mergeReports(reports) {
    const balance = reports['balance_sheet'] || [];
    const income = reports['income_statement'] || [];
    const cash = reports['cash_flow'] || [];

    if (balance.length === 0) {
      return [];
    }

    const merged = balance.map(balanceRow => {
      const mergedRow = { ...balanceRow };

      const matchingIncome = income.find(row => 
        this.mergeKeys.every(key => row[key] === balanceRow[key])
      );

      const matchingCash = cash.find(row =>
        this.mergeKeys.every(key => row[key] === balanceRow[key])
      );

      if (matchingIncome) {
        Object.assign(mergedRow, matchingIncome);
      }

      if (matchingCash) {
        Object.assign(mergedRow, matchingCash);
      }

      return mergedRow;
    });

    return merged;
  }

  /**
   * Bước 5: Kiểm tra tính hợp lệ của dữ liệu
   */
  validateData(data) {
    const validationRules = this.schema.validation_rules || [];

    for (const rule of validationRules) {
      const ruleName = rule.rule;

      if (ruleName === 'non_negative_assets') {
        for (const field of rule.fields) {
          const col = this.validationFields[field];
          if (col) {
            const negativeCount = data.filter(row => {
              const value = parseFloat(row[col]);
              return !isNaN(value) && value < 0;
            }).length;

            if (negativeCount > 0) {
            }
          }
        }
      } else if (ruleName === 'non_negative_liabilities') {
        for (const field of rule.fields) {
          const col = this.validationFields[field];
          if (col) {
            const negativeCount = data.filter(row => {
              const value = parseFloat(row[col]);
              return !isNaN(value) && value < 0;
            }).length;

            if (negativeCount > 0) {
            }
          }
        }
      } else if (ruleName === 'balance_equation') {
        const totalAssetsCol = this.validationFields['total_assets'];
        const totalLiabilitiesCol = this.validationFields['total_liabilities'];
        const equityCol = this.validationFields['owners_equity'];
        const tolerance = rule.tolerance || 0.01;

        const violations = data.filter(row => {
          const assets = parseFloat(row[totalAssetsCol]);
          const liabilities = parseFloat(row[totalLiabilitiesCol]);
          const equity = parseFloat(row[equityCol]);

          if (isNaN(assets) || isNaN(liabilities) || isNaN(equity)) {
            return false;
          }

          const diff = Math.abs(assets - (liabilities + equity));
          return diff > tolerance;
        }).length;

        if (violations > 0) {
        }
      }
    }

    return data;
  }

  /**
   * Chạy toàn bộ pipeline Phase 1
   */
  async standardize(rawDataPath) {

    const rawData = await this.loadRawData(rawDataPath);
    const reports = this.splitByReportType(rawData);
    const cleaned = this.removeEmptyColumns(reports);
    const merged = this.mergeReports(cleaned);
    const validated = this.validateData(merged);

    return validated;
  }

  /**
   * Lấy dữ liệu cho một công ty cụ thể
   */
  getCompanyData(standardizedData, ticker, year, quarter) {
    const filtered = standardizedData.filter(row => {
      const rowTicker = row.ticker || row['﻿ticker']; // Handle BOM
      const rowYear = parseInt(row.yearReport || row.year);
      const rowQuarter = parseInt(row.lengthReport || row.quarter);
      
      return rowTicker === ticker &&
             rowYear === year &&
             rowQuarter === quarter;
    });

    if (filtered.length === 0) {
      throw new Error(`No data found for ${ticker} Q${quarter}/${year}`);
    }

    return filtered[0];
  }
}
